File: tests/csharp/csharp_error_handling/finally_always_runs.cs
// vybe-test: csharp/csharp_error_handling/finally_always_runs
// origin: languages/csharp/tests/csharp/test_csharp_error_handling.rs

string __buf = "";

void __P(string s) {
    __buf = __buf + s + "\n";
}

void __Pr(string s) {
    __buf = __buf + s;
}

// The final WriteLine contributes a trailing newline that the expected line
// vector never carried, so BOTH forms are accepted.
void __Check(string want) {
    if (__buf != want && __buf != want + "\n") {
        Console.WriteLine("FAIL: want [" + want + "] got [" + __buf + "]");
        throw new Exception("assertion failed");
    }
}

try {
    __P(("before").ToString());
    throw new Exception("err");
} catch (Exception e) {
    __P(("caught").ToString());
} finally {
    __P(("always").ToString());
}
__Check("before\ncaught\nalways");

File: tests/dart/errors_advanced/finally_always_runs.dart
// vybe-test: dart/errors_advanced/finally_always_runs
// origin: languages/dart/tests/dart/test_errors_advanced.rs

final StringBuffer __vybeOut = StringBuffer();

void __p(Object? o) {
  __vybeOut.writeln(o);
}

void __check(String want) {
  var got = __vybeOut.toString();
  // `writeln` on the final print contributes a trailing newline that the
  // expected line vector never carried.
  if (got.endsWith('\n')) {
    got = got.substring(0, got.length - 1);
  }
  if (got != want) {
    print('FAIL: want [$want] got [$got]');
    throw Exception('assertion failed');
  }
}

void __vybeMain() {
  try {
    __p('try');
  } finally {
    __p('finally');
  }
}

void main() {
  __vybeMain();
  __check('try\nfinally');
}
